src/model/sample.rs

```diff
@@ -36,40 +36,73 @@ pub fn prob_read_base(read_base: u8, ref_base: u8, base_qual: u8) -> LogProb {
+                &Cigar::SoftClip(_) => {
+                    c = i;
+                    break;
+                },
+                &Cigar::Del(_) => panic!("Cigar String: Unexpected Deletion operation found before any reference match."),
+                &Cigar::Back(_) => panic!("Back: Unsupported Cigar operation (not in SAMv1 spec)."),
+                &Cigar::Pad(_) => panic!("Pad: Unsupported Cigar operation (unclear SAMv1 spec)."),
+                &Cigar::RefSkip(_) => panic!("Cigar String: Unexpected Reference Skip operation found before any reference match."),
+                &Cigar::HardClip(_) => ()
+            }
+
+        }
+        while rpos <= vpos {
+            match &cigar[c] {
                 // potential SNV evidence
-                &Cigar::Match(l) | &Cigar::Diff(l) if contains_start(pos, l) => {
+                &Cigar::Match(l) | &Cigar::Diff(l) | &Cigar::Equal(l) if contains_vpos(rpos, l) => {
+                    qpos += vpos - rpos; // difference between vpos and first position of current cigar operation
                     let read_base = record.seq()[qpos as usize];
                     let base_qual = record.qual()[qpos as usize];
                     let prob_alt = prob_read_base(read_base, base, base_qual);
-                    let prob_ref = prob_read_base(read_base, ref_seq[start as usize], base_qual);
-                    return (prob_ref, prob_alt);
+                    let prob_ref = prob_read_base(read_base, ref_seq[vpos as usize], base_qual);
+                    return Some( (prob_ref, prob_alt) );
                 },
                 // for others, just increase pos and qpos as needed
                 &Cigar::Match(l)   |
-                &Cigar::RefSkip(l) |
-                &Cigar::Equal(l)   |
                 &Cigar::Diff(l)    |
+                &Cigar::Equal(l)   => {
+                    rpos += l;
+                    qpos += l;
+                    c += 1;
+                },
+                &Cigar::SoftClip(l) |
                 &Cigar::Ins(l)  => {
-                    pos += l as i32;
                     qpos += l;
+                    c += 1;
                 },
-                &Cigar::Del(l)  => pos += l as i32,
-                &Cigar::Back(l) => pos -= l as i32,
-                _ => ()
+                &Cigar::RefSkip(l) |
+                &Cigar::Del(l)  => {
+                    rpos += l;
+                    c += 1;
+                },
+                &Cigar::HardClip(_) => return None,
+                &Cigar::Back(_) => panic!("Back: Unsupported Cigar operation (not in SAMv1 spec)."),
+                &Cigar::Pad(_) => panic!("Pad: Unsupported Cigar operation (unclear SAMv1 spec).")
             }
         }
-        panic!("read does not enclose SNV");
+        // if no SNV evidence is found before passing the vpos, must be
+        return None;
```

Comment:
I don't yet understand why the loop is changed like this. Can you point me to the problems with the previous loop?